athena-dynamodb/src/main/java/com/amazonaws/athena/connectors/dynamodb/util/DDBClientBuilder.java

```diff
@@ -0,0 +1,63 @@
+/*-
+ * #%L
+ * athena-dynamodb
+ * %%
+ * Copyright (C) 2019 - 2023 Amazon Web Services
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.amazonaws.athena.connectors.dynamodb.util;
+
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicSessionCredentials;
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClientBuilder;
+import com.amazonaws.services.securitytoken.AWSSecurityTokenService;
+import com.amazonaws.services.securitytoken.AWSSecurityTokenServiceAsyncClientBuilder;
+import com.amazonaws.services.securitytoken.model.AssumeRoleRequest;
+import com.amazonaws.services.securitytoken.model.AssumeRoleResult;
+import com.amazonaws.services.securitytoken.model.Credentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+
+public class DDBClientBuilder
+{
+    private static final String CROSS_ACCOUNT_DYNAMODB_ROLE_ARN = "cross_account_ddb_role_arn";
+    private static final Logger logger = LoggerFactory.getLogger(DDBClientBuilder.class);
+
+    private DDBClientBuilder() {}
+    
+    public static AmazonDynamoDB createDynamoDBClient(Map<String, String> configOptions)
+    {
+        if (configOptions.containsKey(CROSS_ACCOUNT_DYNAMODB_ROLE_ARN)) {
+            logger.debug("Found cross-account role arn to assume for dynamo client.");
+            AWSSecurityTokenService stsClient = AWSSecurityTokenServiceAsyncClientBuilder.standard().build();
+            AssumeRoleRequest assumeRoleRequest = new AssumeRoleRequest()
+                .withRoleArn(configOptions.get(CROSS_ACCOUNT_DYNAMODB_ROLE_ARN))
+                .withRoleSessionName("DynamoDBCrossAccountQuerySession");
+            AssumeRoleResult assumeRoleResult = stsClient.assumeRole(assumeRoleRequest);
+            Credentials credentials = assumeRoleResult.getCredentials();
+            BasicSessionCredentials basicSessionCredentials = new BasicSessionCredentials(credentials.getAccessKeyId(), credentials.getSecretAccessKey(), credentials.getSessionToken());
+            AWSCredentialsProvider credentialsProvider = new AWSStaticCredentialsProvider(basicSessionCredentials);
+            return AmazonDynamoDBClientBuilder
+                .standard()
+                .withCredentials(credentialsProvider)
+                .build();
+        }
```

Comment:
It looks like this code is fairly generic, I think you might be able to structure this in a way where you can just have like a generic "CrossAccountCredentialProvider" in the SDK (which is basically this method that takes in a configOptions map and instead it only returns an AwsCredentialsProvider).

Then the environment variable this generic "CrossAccountCredentialProvider" would be a generic "CrossAccountRoleARN" environment variable.

In the case where that environment variable isn't defined, you can just return: `DefaultAWSCredentialsProviderChain.getInstance()`

Then all the connectors would just call this method blindly and use whatever AwsCredentialsProvider they get back blindly.

Reply:
That's a good idea, let me abstract just the credentials provider out. Callers will just provide the config options map and a name for the role session.